### ISSUE TITLE
Tenant Editable header and footer area :Enhancement

### ISSUE THREAD

**Issue by abomadi** [cloudscribe/cloudscribe#564]
a suggestion to enable add site logo in site settings for each tenant

**Comment by joeaudette** [comment]
adding a data property for site logo url would be easy.

But making a consistent experience for user upload of the image and rendering of an image poses some challenges.
By what rules do we constrain the image size or resize it on upload?
If we do constrain it it could distort the logo or degrade the quality.
If we don't constrain it and for example the image is to be added in layout using that site property it would be very easy for the image to throw off the whole layout.

That is why I've left it for designers to put logos in the theme layout file as needed so the sizing and layout is up to them to handle correctly.

**Comment by CrispinF** [comment]
@joeaudette I'm inclined to agree - it will be very rare that results are acceptable.
But the aim is good... and I wonder whether following the approach now taken for the Blog "About" might be a better way, hydrating a container either above or below the navbar (with the default layout)which would in turn allow for a big site header or a banner image, all controlled from site settings.

**Comment by joeaudette** [comment]
@CrispinF yeah an editable header content area would be a good solution for a logo image with immediate feedback for the user and ability to undo it easily if it blows up the layout.

@abomadi would that be sufficient for you?

**Comment by abomadi** [comment]
I guess so, if it would be available across the all tenant site pages, but i think this will not be available with cloudscribe core, only with simplecontent enabled, other suggestion as u said a data property maybe with a class name property or height and width

**Comment by joeaudette** [comment]
@abomadi no this would be a new feature in cloudscribe core, instead of site.LogoImage it would be site.HeaderContent editable in site settings. Then have to go through all the layouts for included themes and add a partial view for it.

**Comment by abomadi** [comment]
Amazing joe, this will do the needed too.

**Comment by joeaudette** [comment]
ok, so more thinking, probably also should add site.FooterContent at the same time. And probably should add a boolean whether to show the sitename link or not since editing the header one might want to have the site link there or not at all if using a logo, maybe the logo is the site link.

**Comment by CrispinF** [comment]
@joeaudette you took the words right out of my mouth... (Meat Loaf obv)

**Comment by abomadi** [comment]
Brilliant...

**Comment by joeaudette** [comment]
I have implemented this locally and I must say that I'm finding it is not really providing a good user experience. Content added in header and footer does not naturally blend with the design.

If I put the partial above or below the nav it really always renders below the nav due to bootstrap top nav positioning.

Putting it inside the nave right after the sitename link seems the right place for it but then you enter content, ckeditor wants to wrap it in p, select text and use the format bar to change to div helps. But without adding css classes then the text is black and doesn't match the other nav text so you really have to go into source and put class="navbar-text" or class="navbar-brand" to make it look ok.

Even worse experience with images, typically the image is bigger than the navbar height for the design and it makes the navbar grow taller and this obscures body content including the breadcrumbs. Ckeditor wants to wrap images in p, you can change it to div in source view. You have to edit the style attributes on the image, remove width and specify only the height you want so it will shrink without distorting aspect ratio.

So this is all very fiddly and not good UX for non html experts. Having  it in the UI invites people to use it and will lead to bad experience for many people vs not having it there at all. So I'm going to make a theme setting whether to even show that stuff in site settings page and default it to false. It is useable for people who know html well and can manage the quirks of ckeditor. Theme setting will be better than config setting because then it can be enabled for specific sites without enabling it globally.

For example theme setting in appsettings are like this

     "CoreThemeConfig": {
    "DefaultTheme": "default",
    "ThemeSettings": [
      {
        "ThemeName": "default",
        "CoreIconSetId": "fontawesome5x",
        "AdminSideNavExpanded": true
      },
      {
        "ThemeName": "bootstrap4",
        "CoreIconSetId": "fontawesome5x",
        "AdminSideNavExpanded": true
      }
     ]
    },

If you want to enable it globally you could do this:

     "CoreThemeConfig": {
    "DefaultTheme": "default",
    "ThemeSettings": [
      {
        "ThemeName": "default",
        "CoreIconSetId": "fontawesome5x",
        "AdminSideNavExpanded": true, 
        "EnableHeaderFooterEditingInSiteSettings": true
      },
      {
        "ThemeName": "bootstrap4",
        "CoreIconSetId": "fontawesome5x",
        "AdminSideNavExpanded": true
       } 
      ]
    },

It still won't be completely global because the above I added it to default but the more specific theme named "bootstrap4" would not have it unless you add it there too.

In summary though I have implemented this and it will be possible to use it if you want to, it really has the same problems I originally mentioned and why I still think it is best to let the designer add the logo, header and footer markup in the layout and not expose it to users.

**Comment by joeaudette** [comment]
This is now completed and new nugets published, disabled by default as mentioned above

**Comment by joeaudette** [comment]
I think maybe in themes where you create big headers and footers perhaps these new editable areas will be very useful so it is good that it can be enabled per theme. But with the standard bootstrap and bootswatch themes it is only useful for html experts so best to keep it disabled by default.

**Comment by joeaudette** [comment]
Reopening this issue because I have some ideas. 

Thinking that I will go ahead and add a LogoUrl property soon and make something to drag drop an image to upload and set as logo url. Whether it is shown on the edit page and in the header can also be based on a theme setting.

I don’t think it would work well for the currently included themes, but I have an idea if I modify those I might be able to make it work. Maybe by adding a specific height to the navbar like 150px then could set the logo height to some percent and that might make it behave reasonably well.

Need to do some experimenting.

**Comment by abomadi** [comment]
Im not sure if this gonna help, but here is an example saying logo will be included to bootstrap 4.2.1 
https://startbootstrap.com/snippets/navbar-logo/

**Comment by abomadi** [comment]
@joeaudette i was testing the new functionality, it seems the <p> tag from the editor is running the experince, my suggestion is as u said adding a property to site logo as a partial view that only inserts logo if the property has value.. and i tested this code that works well too
```
<nav class="navbar main-nav navbar-expand-lg navbar-dark" aria-label="Main navigation">
<a class="navbar-brand " href="#">
                    <img src="https://test.com/Logo.png" height="30" class="img-responsive" alt="Logo">
                </a>
```
also height and width maybe could be edited with the logo property.
This will insert the logo perfectly, however any advanced logo customized it could be done by designers by overriding SiteHeaderPartial.

**Comment by joeaudette** [comment]
Ok, I've published updated nugets with log image in site settings.

To make it work with the included themes I set only the img height as 100% I don't set width, so it keeps the correct image aspect ratio.
But the containing element of the img is an a element so I have to set a height on it and on the nav div to constrain the image to fit the theme. I added a new setting for that in theme settings:

    "CoreThemeConfig": {
    "DefaultTheme": "default",
    "ThemeSettings": [
      {
        "ThemeName": "default",
        "CoreIconSetId": "fontawesome5x",
        "AdminSideNavExpanded": true,
        "EnableHeaderFooterEditingInSiteSettings":false,
        "NavbarHeightInPixels":60
      },
      {
        "ThemeName": "bootstrap4",
        "CoreIconSetId": "fontawesome5x",
        "AdminSideNavExpanded": true,
        "EnableHeaderFooterEditingInSiteSettings":false,
        "NavbarHeightInPixels":60
      }
       ]
     },

I used 60 because that works reasonable with the included themes.

So the themes had to be updated to inject themeHelper in layout and get the themesettings, and themesettings has some helper functions used to set style on the nav, and the a that wraps the image  

    <nav class="navbar navbar-expand-lg navbar-dark fixed-top bg-dark" style="@themeSetings.GetNavbarStyle()">
            @if (!string.IsNullOrWhiteSpace(Tenant.LogoUrl))
            {
                <a style="@themeSetings.GetNavbarStyle()" href="~/@(Tenant?.SiteFolderName ?? "")" ><img src="@Tenant.LogoUrl" style="height:100%;" class="mr-1" /></a>  
            }
            @if (Tenant.ShowSiteNameLink)
            {
                <a href="~/@(Tenant?.SiteFolderName ?? "")" class="navbar-brand">@Tenant.SiteName </a>
            }
            <partial name="SiteHeaderPartial" />
            <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarNavDropdown" aria-controls="navbarNavDropdown" aria-expanded="false" aria-label="Toggle navigation">
                <span class="navbar-toggler-icon"></span>
            </button>

            <div class="collapse navbar-collapse" id="navbarNavDropdown">
                @await Component.InvokeAsync("Navigation", new { viewName = "Bootstrap4TopNavWithDropdowns", filterName = NamedNavigationFilters.TopNav, startingNodeKey = "" })
                <partial name="_LoginPartialWithDropdowns" />
            </div>
        </nav>

note the style attributes.

Also because the topnav is fixed position we need padding on the body equivalent to the nav height otherwise content including breadcrumbs would be obscured by the header. So you will also see the body element has been updated like this:

    <body class="d-flex flex-column @ViewData["BodyClass"]" style="@themeSetings.GetBodyStyle()">

**Comment by joeaudette** [comment]
just noted in copy pasting the code I made some typos in the variable names in the layouts. I will fix those when I get a chance but there are 22 included layout files so its kind of a pain to fix and noone but developers will see it anyway.

ie I have

    @inject ICoreThemeHelper themHelper

which should have been

    @inject ICoreThemeHelper themeHelper

and

    var themeSetings = themHelper.GetThemeSettings();

which should have been 

    var themeSettings = themHelper.GetThemeSettings();

I'm the king of typos!

**Comment by joeaudette** [comment]
anyway get the updated project template, try it out and let me know what you think.

**Comment by joeaudette** [comment]
just fixed the typos with find/replace and published updated templates again

**Comment by joeaudette** [comment]
@all fyi just published update project template v 1.65

I went through each of the bootswatch themes and fixed some issues when using a site logo. Also noticed some problems with the responsive hamburger menu and fixed those.

**Comment by joeaudette** [comment]
@all fyi, I finally got something working reasonable for all the included themes to have the editable header and footer. Mainly I added a header area above the navbar and moved the site title, logo, and editable area there. I also did some tweaking of the footer styles to work better on small screens and to accommodate the editable area.

I've published an updated project template with the latest theme updates. You should see the difference if creating new projects with the new template.
https://marketplace.visualstudio.com/items?itemName=joeaudette.cloudscribeProjectTemplate
